File: sdk/server/python/aurix_server/_http.py
"""Hand-written transport under the generated client: auth headers, JSON/raw requests,
timeouts, bounded retries. Standard library only (``urllib``)."""

from __future__ import annotations

import json
import random
import socket
import ssl
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from typing import Any, Dict, List, Mapping, Optional, Tuple

from .errors import AurixError, AurixNetworkError

SDK_VERSION = "1.2.0"
_IDEMPOTENT = frozenset({"GET", "HEAD", "OPTIONS", "PUT", "DELETE"})
_RETRY_STATUSES = frozenset({502, 503, 504})


@dataclass
class Credentials:
    """Exactly one of these is normally set. Precedence: api_key, admin_token, player_token, bootstrap_token."""

    api_key: Optional[str] = None
    admin_token: Optional[str] = None
    player_token: Optional[str] = None
    bootstrap_token: Optional[str] = None

    def headers(self) -> Dict[str, str]:
        if self.api_key:
            return {"X-API-Key": self.api_key}
        if self.admin_token:
            return {"Authorization": f"Bearer {self.admin_token}"}
        if self.player_token:
            return {"Authorization": f"Bearer {self.player_token}"}
        if self.bootstrap_token:
            return {"X-Bootstrap-Token": self.bootstrap_token}
        return {}


@dataclass
class RequestOptions:
    """Per-call overrides."""

    headers: Dict[str, str] = field(default_factory=dict)
    timeout: Optional[float] = None
    auth: Optional[Credentials] = None


@dataclass
class RawResponse:
    """Undecoded response for binary / CSV / SRT / VTT operations."""

    status: int
    content_type: str
    headers: Dict[str, str]
    body: bytes

    def text(self, encoding: str = "utf-8") -> str:
        return self.body.decode(encoding)

    def json(self) -> Any:
        return json.loads(self.body)


def path_segment(value: Any) -> str:
    return urllib.parse.quote(str(value), safe="")


def _query_items(query: Optional[Mapping[str, Any]]) -> List[Tuple[str, str]]:
    out: List[Tuple[str, str]] = []
    if not query:
        return out
    for key, value in query.items():
        if value is None:
            continue
        if isinstance(value, (list, tuple)):
            out.extend((key, _qs(v)) for v in value if v is not None)
        else:
            out.append((key, _qs(value)))
    return out


def _qs(value: Any) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def retry_after_seconds(headers: Mapping[str, str]) -> Optional[float]:
    for k, v in headers.items():
        if k.lower() == "retry-after":
            try:
                return max(0.0, float(v))
            except ValueError:
                return None
    return None


class BaseClient:
    """Shared behaviour of the generated ``AurixClient``.

    ``base_url`` is the node's HTTP origin (``https://voice.example.com``). Authenticate with
    exactly one credential: ``api_key`` (game backends), ``admin_token`` (operators),
    ``player_token`` (acting on behalf of a player) or ``bootstrap_token`` (first admin).
    """

    def __init__(
        self,
        base_url: str,
        *,
        api_key: Optional[str] = None,
        admin_token: Optional[str] = None,
        player_token: Optional[str] = None,
        bootstrap_token: Optional[str] = None,
        timeout: float = 15.0,
        max_retries: int = 2,
        max_backoff: float = 5.0,
        headers: Optional[Mapping[str, str]] = None,
        user_agent: Optional[str] = None,
        ssl_context: Optional[ssl.SSLContext] = None,
        opener: Optional[urllib.request.OpenerDirector] = None,
    ):
        self.base_url = base_url.rstrip("/")
        self.credentials = Credentials(api_key, admin_token, player_token, bootstrap_token)
        self.timeout = timeout
        self.max_retries = max(0, max_retries)
        self.max_backoff = max_backoff
        self.default_headers = dict(headers or {})
        self.user_agent = user_agent or f"aurix-server-sdk-python/{SDK_VERSION}"
        https = urllib.request.HTTPSHandler(context=ssl_context) if ssl_context else urllib.request.HTTPSHandler()
        self._opener = opener or urllib.request.build_opener(https)

    def url(self, path: str, query: Optional[Mapping[str, Any]] = None) -> str:
        url = self.base_url + path
        items = list(_query_items(query))
        if items:
            url += "?" + urllib.parse.urlencode(items)
        return url

    def _headers(self, options: Optional[RequestOptions], has_body: bool, accept: str) -> Dict[str, str]:
        h: Dict[str, str] = {"Accept": accept, "User-Agent": self.user_agent}
        h.update(self.default_headers)
        h.update((options.auth if options and options.auth else self.credentials).headers())
        if has_body:
            h["Content-Type"] = "application/json"
        if options:
            h.update(options.headers)
        return h

    def _json(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: Optional[Mapping[str, Any]] = None,
        options: Optional[RequestOptions] = None,
    ) -> Any:
        raw = self._raw(method, path, body=body, query=query, options=options, accept="application/json")
        if raw.status == 204 or not raw.body:
            return None
        if not raw.content_type.lower().startswith("application/json"):
            raise AurixError(
                status=raw.status,
                code="unexpected_content_type",
                message=f"expected application/json, got {raw.content_type or 'no content type'}; use the *_raw variant",
                method=method,
                path=path,
                body=raw.body,
            )
        return json.loads(raw.body)

    def _raw(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: Optional[Mapping[str, Any]] = None,
        options: Optional[RequestOptions] = None,
        accept: str = "*/*",
    ) -> RawResponse:
        method = method.upper()
        url = self.url(path, query)
        data = json.dumps(body).encode("utf-8") if body is not None else None
        headers = self._headers(options, data is not None, accept)
        timeout = options.timeout if options and options.timeout is not None else self.timeout
        attempt = 0
        while True:
            attempt += 1
            try:
                req = urllib.request.Request(url, data=data, method=method, headers=headers)
                with self._opener.open(req, timeout=timeout) as resp:
                    status = resp.status
                    resp_headers = {k: v for k, v in resp.headers.items()}
                    content = resp.read()
            except urllib.error.HTTPError as e:
                status = e.code
                resp_headers = {k: v for k, v in e.headers.items()}
                content = e.read()
            except (urllib.error.URLError, socket.timeout, TimeoutError, ConnectionError, ssl.SSLError, OSError) as e:
                if attempt <= self.max_retries and method in _IDEMPOTENT:
                    time.sleep(self._backoff(attempt))
                    continue
                raise AurixNetworkError(str(getattr(e, "reason", e) or e), method, path, e) from e

            content_type = next((v for k, v in resp_headers.items() if k.lower() == "content-type"), "")
            if 200 <= status < 300:
                return RawResponse(status, content_type.split(";")[0].strip(), resp_headers, content)
            retryable = status == 429 or (status in _RETRY_STATUSES and method in _IDEMPOTENT)
            if retryable and attempt <= self.max_retries:
                delay = retry_after_seconds(resp_headers)
                time.sleep(self._backoff(attempt) if delay is None else min(delay, self.max_backoff))
                continue
            raise AurixError.from_response(status, content_type, content, resp_headers, method, path)

    def _backoff(self, attempt: int) -> float:
        base = min(self.max_backoff, 0.2 * float(2 ** (attempt - 1)))
        return float(base * (0.5 + random.random() / 2))

File: sdk/server/node/src/http.ts
/**
 * Hand-written transport under the generated `AurixClient`: authentication headers, JSON
 * encoding, query strings, timeouts, retries with `Retry-After`, and error mapping.
 */

import { AurixError, AurixNetworkError } from "./errors.js";
import { eventStream, type EventStreamOptions, type SseEvent } from "./sse.js";

/** Exactly one credential should be set; the API key is the normal server-side choice. */
export interface AurixCredentials {
  /** Application API key (`aurx_...`), sent as `X-API-Key`. Never ship it to game clients. */
  apiKey?: string;
  /** Admin JWT from `POST /admin/login` / OIDC, sent as `Authorization: Bearer`. */
  adminToken?: string;
  /** Player session JWT (for the `/v1/me/*` endpoints). */
  playerToken?: string;
  /** `auth.bootstrap_token`, only for `POST /admin/setup`. */
  bootstrapToken?: string;
}

export interface AurixClientOptions extends AurixCredentials {
  /** Control-plane base URL, e.g. `https://voice.example.com` (no trailing slash needed). */
  baseUrl: string;
  /** Per-request timeout in milliseconds (default 10 000). */
  timeoutMs?: number;
  /**
   * Retries for idempotent requests (GET/PUT/DELETE) on network errors, 502/503/504, and for
   * any method on 429 (the request was not processed). Default 2; `0` disables.
   */
  maxRetries?: number;
  /** Upper bound for a single backoff / `Retry-After` wait (default 5 000 ms). */
  maxBackoffMs?: number;
  /** `fetch` implementation (default: global `fetch`, Node ≥ 18). */
  fetch?: typeof fetch;
  /** Extra headers on every request. */
  headers?: Record<string, string>;
  userAgent?: string;
}

export interface RequestOptions {
  signal?: AbortSignal;
  headers?: Record<string, string>;
  timeoutMs?: number;
  /** Override the client credentials for this call only (e.g. a player token on `/v1/me/*`). */
  auth?: AurixCredentials;
}

export interface RawResponse {
  status: number;
  contentType: string;
  headers: Headers;
  body: Uint8Array;
  /** Decodes `body` as UTF-8. */
  text(): string;
}

interface CallOptions extends RequestOptions {
  body?: unknown;
  query?: object | undefined;
}

export const SDK_VERSION = "1.5.0";
const RETRY_METHODS = new Set(["GET", "PUT", "DELETE", "HEAD"]);
const RETRY_STATUSES = new Set([502, 503, 504]);

export class AurixHttp {
  readonly baseUrl: string;
  /** @internal */
  readonly opts: AurixClientOptions;
  /** @internal */
  readonly fetchImpl: typeof fetch;

  constructor(options: AurixClientOptions) {
    if (!options.baseUrl) throw new Error("AurixClient: baseUrl is required");
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.opts = options;
    const f = options.fetch ?? globalThis.fetch;
    if (typeof f !== "function") throw new Error("AurixClient: no fetch available; pass options.fetch");
    // Browsers reject `fetch` called with a foreign `this` ("Illegal invocation").
    this.fetchImpl = options.fetch ? f : (input, init) => globalThis.fetch(input, init);
  }

  /** Builds the absolute URL for `path` + `query` (nullish values are skipped, arrays repeat). */
  url(path: string, query?: object): string {
    const u = new URL(this.baseUrl + path);
    if (query) {
      for (const [k, v] of Object.entries(query as Record<string, unknown>)) {
        if (v === undefined || v === null) continue;
        if (Array.isArray(v)) {
          for (const item of v) u.searchParams.append(k, String(item));
        } else {
          u.searchParams.set(k, String(v));
        }
      }
    }
    return u.toString();
  }

  /** @internal */
  authHeaders(auth: AurixCredentials): Record<string, string> {
    const h: Record<string, string> = {};
    if (auth.apiKey) h["X-API-Key"] = auth.apiKey;
    else if (auth.adminToken) h["Authorization"] = `Bearer ${auth.adminToken}`;
    else if (auth.playerToken) h["Authorization"] = `Bearer ${auth.playerToken}`;
    if (auth.bootstrapToken) h["X-Bootstrap-Token"] = auth.bootstrapToken;
    return h;
  }

  /**
   * `GET /v1/events` as an async iterator (auto-reconnect, `Last-Event-ID` resume). See `eventStream`.
   */
  events(options?: EventStreamOptions): AsyncGenerator<SseEvent, void, undefined> {
    return eventStream(this, options);
  }

  /** JSON request; resolves to `undefined` on an empty body. */
  async json<T>(method: string, path: string, call: CallOptions): Promise<T> {
    const res = await this.send(method, path, call);
    if (res.status === 204 || res.body.byteLength === 0) return undefined as T;
    if (!isJson(res.contentType)) {
      throw new AurixError({
        status: res.status,
        code: "unexpected_content_type",
        message: `expected application/json but got ${res.contentType || "no content type"}; use the Raw variant of this call`,
        method,
        path,
      });
    }
    return JSON.parse(res.text()) as T;
  }

  /** Raw request (binary / CSV / SRT / VTT bodies). */
  raw(method: string, path: string, call: CallOptions): Promise<RawResponse> {
    return this.send(method, path, call);
  }

  private async send(method: string, path: string, call: CallOptions): Promise<RawResponse> {
    const url = this.url(path, call.query);
    const headers: Record<string, string> = {
      Accept: "application/json, */*;q=0.5",
      "User-Agent": this.opts.userAgent ?? `aurix-server-sdk-node/${SDK_VERSION}`,
      ...this.opts.headers,
      ...this.authHeaders(call.auth ?? this.opts),
      ...call.headers,
    };
    let body: string | undefined;
    if (call.body !== undefined) {
      headers["Content-Type"] = "application/json";
      body = JSON.stringify(call.body);
    }
    const maxRetries = this.opts.maxRetries ?? 2;
    const timeoutMs = call.timeoutMs ?? this.opts.timeoutMs ?? 10_000;
    const maxBackoff = this.opts.maxBackoffMs ?? 5_000;
    const upper = method.toUpperCase();
    for (let attempt = 0; ; attempt++) {
      const controller = new AbortController();
      const onAbort = () => controller.abort(call.signal?.reason);
      if (call.signal) {
        if (call.signal.aborted) throw new AurixNetworkError("request aborted", method, path);
        call.signal.addEventListener("abort", onAbort, { once: true });
      }
      const timer = setTimeout(() => controller.abort(new Error(`timeout after ${timeoutMs} ms`)), timeoutMs);
      let res: Response;
      try {
        res = await this.fetchImpl(url, { method: upper, headers, body: body ?? null, signal: controller.signal });
      } catch (err) {
        clearTimeout(timer);
        call.signal?.removeEventListener("abort", onAbort);
        if (call.signal?.aborted) throw new AurixNetworkError("request aborted", method, path, err);
        if (attempt < maxRetries && RETRY_METHODS.has(upper)) {
          await sleep(backoff(attempt, maxBackoff));
          continue;
        }
        throw new AurixNetworkError(err instanceof Error ? err.message : String(err), method, path, err);
      }
      let bytes: Uint8Array;
      try {
        bytes = new Uint8Array(await res.arrayBuffer());
      } finally {
        clearTimeout(timer);
        call.signal?.removeEventListener("abort", onAbort);
      }
      const contentType = res.headers.get("content-type") ?? "";
      if (res.ok) {
        return { status: res.status, contentType, headers: res.headers, body: bytes, text: () => decode(bytes) };
      }
      const retryable = res.status === 429 || (RETRY_METHODS.has(upper) && RETRY_STATUSES.has(res.status));
      if (retryable && attempt < maxRetries) {
        await sleep(retryAfterMs(res.headers.get("retry-after")) ?? backoff(attempt, maxBackoff), maxBackoff);
        continue;
      }
      throw AurixError.fromResponse(res.status, contentType, decode(bytes), res.headers, method, path);
    }
  }
}

function isJson(contentType: string): boolean {
  const ct = contentType.toLowerCase();
  return ct.startsWith("application/json") || ct.includes("+json");
}

function decode(bytes: Uint8Array): string {
  return new TextDecoder().decode(bytes);
}

function backoff(attempt: number, max: number): number {
  const base = Math.min(max, 200 * 2 ** attempt);
  return Math.round(base / 2 + Math.random() * (base / 2));
}

export function retryAfterMs(header: string | null): number | undefined {
  if (!header) return undefined;
  const secs = Number(header);
  if (Number.isFinite(secs)) return Math.max(0, secs * 1000);
  const at = Date.parse(header);
  return Number.isNaN(at) ? undefined : Math.max(0, at - Date.now());
}

function sleep(ms: number, max = ms): Promise<void> {
  return new Promise((r) => setTimeout(r, Math.min(ms, max)));
}
